rfc/text/advanced/ap_authentication_scram.md

```diff
@@ -0,0 +1,387 @@
+4. `iterations|integer`: The execution time cost factor to use for generating the `SaltedPassword` hash. This value is stored with each user record in the authentication database.
+5. `memory|integer`: The memory cost factor to use for generating the `SaltedPassword` hash. This is only used by the Argon2 key derivation function, where it is stored with each user record in the authentication database.
+
+The client MUST respond with an ABORT message if `CHALLENGE.Details.nonce` does not begin with the client nonce sent in `HELLO.Details.nonce`.
+
+The client SHOULD respond with an ABORT message if it detects that the cost parameters are unusually low. Such low-cost parameters could be the result of a rogue server attempting to obtain a weak password hash that can be easily cracked. What constitutes unusually low parameters is implementation-specific and is not covered by this document.
+
+
+###### Final Client Authentication Message
+
+Upon receiving a valid `CHALLENGE` message, the client SHALL respond with an `AUTHENTICATE` message:
+
+{align="left"}
+```javascript
+    [5, "dHzbZapWIk4jUhN+Ute9ytag9zjfMHgsqmmiz7AndVQ=",
+        {
+            "nonce": "egVDf3DMJh0=SBmkFIh7sSo=",
+            "channel_binding": null,
+            "cbind_data": null
+        }
+    ]
+```
+where:
+
+1. `Signature|string` argument: The base64-encoded `ClientProof`, computed as described in the [SCRAM-Algorithms](#scram-algorithms) section.
+2. `nonce|string`: The concatenated client-server nonce from the previous `CHALLENGE` message.
+3. `channel_binding|string`: Optional string containing the channel binding type that was sent in the original `HELLO` message.
+4. `cbind_data|string`: Optional base64-encoded channel binding data. MUST be present if and only if `channel_binding` is not `null`. The format of the binding data is dependent on the binding type. For `tls-server-end-point`, this is a hash of the TLS server's certificate, computed as described in [RFC5929, section 4.1](https://tools.ietf.org/html/rfc5929#section-4.1).
+
+Upon receiving the `AUTHENTICATE` message, the server SHALL then check that:
+
+* The `AUTHENTICATE` message was received in due time.
+* The `ClientProof` passed via the `Signature|string` argument is validated against the `StoredKey` and `ServerKey` stored in the authentication database. See [SCRAM Algorithms](#scram-algorithms).
+* `nonce` matches the one previously sent via `CHALLENGE`.
+* The `channel_binding` matches the one sent in the `HELLO` message.
+* The `cbind_data`, if applicable, is valid.
+
+
+###### Final Server Authentication Message - Success
+
+If the authentication succeeds, the server SHALL finally respond with a `WELCOME` message:
+
+{align="left"}
+```javascript
+    [2, 3251278072152162,
+        {
+            "authid": "user",
+            "authrole": "frontend",
+            "authmethod": "wamp-scram",
+            "authprovider": "static",
+            "roles": ...,
+            "authextra":
```

Comment:
yes, if the wamp-scram here would be part of a future multi-factor scheme, then we would need a list of dicts for WELCOME.authextra here